Add support for Vert.x Web sessions

Ladicek opened the pull request:
The individual commits are:

- Infinispan Client: improve dev services logging
- Infinispan Client: add runtime SPI and deployment SPI modules
- Redis Client: add runtime SPI and deployment SPI modules
- Add support for Vert.x Web sessions

The last commit being the biggest obviously. Tagging @karesti for review of the 1st and 2nd commit, and @cescoffier for review of the 3rd commit.

sberyozkin commented:
@Ladicek Is it related to https://vertx.io/docs/vertx-web/java/#_handling_sessions ?

@michalvavrik completed the PR providing support for OIDC session data (tokens) stored in DB - the concept is the same.
Where does Vert.x session handler save the data ? We may as well optionally support it as an implementation of OIDC `TokenStateManager` if Vert.x Web session can be activated with some single or a a very few properties

Ladicek commented:
Hi @sberyozkin, I've been meaning to reply to your email about stateful OIDC token management after I get this PR up, but you beat me to it :-) For one, sessions are just fine, and two, I also thought you could just use sessions. We can discuss in that thread.

But yes, this is indeed about adding https://vertx.io/docs/vertx-web/java/#_handling_sessions to Quarkus. I aim to support all the "reasonable" session stores that Vert.x supports -- in-memory (local), clustered, Redis, Infinispan. I _don't_ want to support cookieless sessions, and I don't want to support the cookie session store. It seems to work rather nicely, but I've only done some basic manual tests so far.

Also, I'd love to know your (@sberyozkin @michalvavrik) opinion on the default value of the `SameSite` attribute of the session cookie. In Vert.x, the attribute is not present by default and you have to configure it explicitly. I'm following that approach for now, but I thought maybe we could use a more reasonable default (`Lax` or `Strict`). WDYT?

michalvavrik commented:
I don't want to pollute your fantastic PR with OIDC token state stuff, but I'm all in. Let's discuss it in Quarkus DEV group when you find time. I'll wait. Using sessions, especially with Redis / Infinispan makes total sense. Thanks!

> Also, I'd love to know your (@sberyozkin @michalvavrik) opinion on the default value of the `SameSite` attribute of the session cookie. In Vert.x, the attribute is not present by default and you have to configure it explicitly. 

I thought when samesite cookie is not present, browsers mostly treat is as `lax`. Definitely +1 to explicitly specify it.

> I'm following that approach for now, but I thought maybe we could use a more reasonable default (`Lax` or `Strict`). WDYT?

TBH I don't know what is reasonable default. FYI OIDC extension does for so called session cookie.

cescoffier commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/SessionsInfinispanConfig.java:
```diff
@@ -0,0 +1,27 @@
+     * Name of the Infinispan cache used to store session data. If it does not exist, it is created
+     * automatically from Infinispan's default template {@code DIST_SYNC}.
+     */
+    @ConfigItem(defaultValue = "vertx-web.sessions")
```
I'm wondering if we need to make it vert.x-web, and more quarkus.sessions? No strong opinions.

Ladicek replied:
I mostly just followed Vert.x Web defaults, the only place I didn't was the name of the session cookie. It probably makes sense to change this to `quarkus.sessions`.

cescoffier commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:
```diff
@@ -346,6 +357,36 @@ public void mountFrameworkRouter(RuntimeValue<Router> mainRouter, RuntimeValue<R
+                    httpConfiguration.getValue().sessions.inMemory.mapName,
+                    httpConfiguration.getValue().sessions.inMemory.retryTimeout.toMillis()));
+        } else {
+            // TODO maybe make reaper interval also configurable?
```
How long is the default?

Ladicek replied:
1 second.

cescoffier commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:
```diff
@@ -413,6 +454,23 @@ public void handle(RoutingContext routingContext) {
+                    .setCookieSameSite(sessions.cookieSameSite.orElse(null))
+                    .setCookieMaxAge(sessions.cookieMaxAge.map(Duration::toMillis).orElse(-1L));
+            // TODO verify if this is the correct router to which the session handler should be installed
+            httpRouteRouter.route(path).handler(sessionHandler);
```
I would say yes. That's the main "public" router. The question, however, is the order of the route.

Ladicek replied:
Yes, good point. I need to explore more around this, that's why I added these TODOs.

cescoffier reviewed (changes requested):
Looks great! I made a few comments. Nothing critical.

Ladicek commented on extensions/infinispan-client/sessions/runtime/src/main/resources/META-INF/quarkus-extension.yaml:
```diff
@@ -0,0 +1,12 @@
+  guide: "https://quarkus.io/guides/http-reference#vertx-web-sessions"
+  categories:
+    - "web"
+  status: "stable"
```
I just copied the status from the Infinispan extension itself without thinking too much about it, but I agree we should downgrade for a while. Will change to `preview`.

Ladicek commented on extensions/redis-client/sessions/runtime/src/main/java/io/quarkus/redis/sessions/runtime/RedisSessionsRecorder.java:
```diff
@@ -0,0 +1,41 @@
+                Annotation qualifier = clientName != null
+                        ? RedisClientName.Literal.of(clientName)
+                        : Default.Literal.INSTANCE;
+                Instance<Redis> bean = Arc.container().select(Redis.class, qualifier);
```
I'm producing the build item to make sure the bean is created: https://github.com/quarkusio/quarkus/pull/36310/files#diff-df2980ec28bea26d72768a6d53aeff1343c4c879f0d5d129b0eb56520c33c550R31

mkouba commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/CurrentVertxRequest.java:
```diff
@@ -17,6 +21,16 @@ public RoutingContext getCurrent() {
+        // but that would break when injected into `@Singleton` or `@ApplicationScoped` beans;
+        // instead, when there's no session, we return a dummy implementation that always throws
+        Session result = current.session();
+        return result != null ? result : DummySession.INSTANCE;
```
Hm, what's the difference compared to throwing an exception directly in the producer method? I mean there's no way to use the dummy session at all. And the producer method is called when the injected proxy is actually used. So if I have `@Inject Session session` then `session.id()` would fail in both cases :shrug:.

Ladicek replied:
That is probably a good point. I need to think more about it. I admit I wrote this code on Friday afternoon, so it isn't my best for sure :-)

Ladicek replied:
Fixed.

mkouba commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/VertxHttpProcessor.java:
```diff
@@ -186,6 +191,17 @@ public KubernetesPortBuildItem kubernetesForManagement(
 
+    @BuildStep
+    public void sessionStoreRequest(HttpBuildTimeConfig config, BuildProducer<SessionStoreRequestBuildItem> request) {
+        if (config.sessions.mode == SessionsBuildTimeConfig.SessionsMode.REDIS) {
```
I wonder if we should check the [available capabilities](io.quarkus.redis) here; i.e. consume the `io.quarkus.deployment.Capabilities` and then `capabilites.isPresent(Capability.REDIS_CLIENT)`...

Ladicek replied:
I originally had that, but the request/response protocol here (which I don't really like, but I don't have a better way at the moment) subsumes that. If the Redis extension isn't present, it won't reply, and we'll get an error.

geoand commented:
@Ladicek sorry, haven't had time to review yet, but I would like to ask about your first question (regarding brining in the deps). Have you had that answered yet? 
If not, I have a suggestion for you.

sberyozkin commented:
@Ladicek Hey, given that your PR is relevant to what OIDC code flow does, I'll briefly reply to some comments

>For one, sessions are just fine, and two, I also thought you could just use sessions. We can discuss in that thread.

Default OIDC session management with encrypted cookies is simplest and most effective and it is recommended, but as I mentioned, what Michal did with providing an alternative option for users to get the session state saved in DB is conceptually exactly identical to what Vert.x Web Sessions offer - the actionable state is saved on the server, the cookie keeps some link to it only.

>I aim to support all the "reasonable" session stores that Vert.x supports -- in-memory (local), clustered, Redis, Infinispan.

For OIDC, `in-memory` strategy would not be useful. OIDC session cookie outlasting the `in-memory` data due to the pod restart/scaling will point to nothing; may be for testing only.

As far as the `clustered, Redis, Infinispan` options are concerned, it is fine, for those OIDC users who don't trust even encrypted cookies and don't work with DBs and thus can't select the option done by Michal where tokens are saved in DB, one of these advanced Vertx Web Session options would be fine :+1:  

Can you please look at https://github.com/quarkusio/quarkus/blob/main/extensions/oidc-db-token-state-manager/runtime/src/main/java/io/quarkus/oidc/db/token/state/manager/runtime/OidcDbTokenStateManager.java ?

For us in OIDC be able to select one of these Vert.x Web Session options, we'd need to sync OIDC `TokenStateManager` API and Vert.x Web Session API that your PR represents, similarly to how Michal did it for linking `TokenStateManager` API with Reactive DB client API

Thanks

Ladicek commented:
> @Ladicek sorry, haven't had time to review yet, but I would like to ask about your first question (regarding brining in the deps). Have you had that answered yet? If not, I have a suggestion for you.

I'd love to hear anything you have!

geoand commented:
A while back @aloubyansky introduced [conditional extension dependencies](https://quarkus.io/guides/conditional-extension-dependencies).
Although the feature is not as widely used as it should be, I have personally used it a few times (I can get references to the source code if you like) and it works amazingly well.
IIUC, it would cover your use case here.

Ladicek commented:
Yeah, so I looked into conditional dependencies and found that I can't really use them, or at least not easily. Or I don't understand conditional dependencies properly. The present situation is (described using Redis, but it's the same for Infinispan):

To be able to store sessions to Redis, one needs to depend on `quarkus-redis-client-sessions`. This module provides a Vert.x Web `SessionStore`, so it needs to depend on Vert.x Web. It also needs to depend on `quarkus-redis-client`, because it needs the CDI qualifier defined there to be able to lookup the Redis client at runtime.

If I wanted to use conditional dependencies, the `quarkus-redis-client-sessions` module would be the "special" one. I'd like it conditionally brought in when both `quarkus-vertx-http` and `quarkus-redis-client` are present. Now, I can't make `quarkus-redis-client` have an optional dependency on `quarkus-redis-client-sessions`, because that would lead to circular dependency. I'm not sure if I can make `quarkus-vertx-http` have an optional dependency on `quarkus-redis-client-sessions`, it seems ... strange.

Maybe I could use [this](https://quarkus.io/guides/conditional-extension-dependencies#declaring-a-conditional-dependency-in-the-quarkus-extension-descriptor) to avoid the dependency circularity? I didn't find any actual usage of that, so I'm not sure how that's supposed to work. I'll try, I guess...

geoand commented:
>  To be able to store sessions to Redis, one needs to depend on quarkus-redis-client-sessions. This module provides a Vert.x Web SessionStore, so it needs to depend on Vert.x Web. It also needs to depend on quarkus-redis-client, because it needs the CDI qualifier defined there to be able to lookup the Redis client at runtime.

> If I wanted to use conditional dependencies, the quarkus-redis-client-sessions module would be the "special" one. I'd like it conditionally brought in when both quarkus-vertx-http and quarkus-redis-client are present. Now, I can't make quarkus-redis-client have an optional dependency on quarkus-redis-client-sessions, because that would lead to circular dependency. I'm not sure if I can make quarkus-vertx-http have an optional dependency on quarkus-redis-client-sessions, it seems ... strange.

Is there any way to move stuff into "common" type dependencies and then wire stuff up? 
I think that's what I had to do with Spring Web and RESTEasy Reactive - it was not trivial nor straightforward.

> Maybe I could use [this](https://quarkus.io/guides/conditional-extension-dependencies#declaring-a-conditional-dependency-in-the-quarkus-extension-descriptor) to avoid the dependency circularity? I didn't find any actual usage of that, so I'm not sure how that's supposed to work. I'll try, I guess...

Okay, I indeed have not had to use that.

However, I _think_ the situation you describe is similar to

FroMage commented:
> I don't want to support the cookie session store.

Why not? This is really the easiest to use for our users, because it requires no server dependency (unlike redis/etc) and allows for backend scaling unlike in-memory sessions.

aloubyansky commented:
> Maybe I could use [this](https://quarkus.io/guides/conditional-extension-dependencies#declaring-a-conditional-dependency-in-the-quarkus-extension-descriptor) to avoid the dependency circularity? I didn't find any actual usage of that, so I'm not sure how that's supposed to work. I'll try, I guess...

I think you should be able to do that. Checking for a condition of Maven optional dependencies is a convenience, since, I was thinking at the time, typically they would be optional dependencies anyway.

Ladicek commented:
Because:

- it is not encrypted, so you can't put sensitive information into it (and you probably shouldn't even if it was)
- it is limited by the nature of cookies and the overhead of Base64 encoding (maximum cookie size is typically about 4 KB, Base64 has roughly 33% overhead, plus there's a few bytes of the signature and the cookie name, so you can safely store, let's say, 2.5 KB of data)
- it increases the size of data you need to transmit over network (potentially significantly, depending on the session data lifecycle)

FroMage commented:
I've played video games with 2.5K ;)
In many cases I don't need my session data to be encrypted, just signed. Things like username, or even JWT token, don't need to be encrypted.

Ladicek commented:
> Is there any way to move stuff into "common" type dependencies and then wire stuff up?

I'm thinking maybe if the qualifier defined by `quarkus-redis-client` was in something like `quarkus-redis-client-common`, that could prevent the circular dependency. Alternatively, I saw `quarkus-infinispan-client` emits build items that allow access to the clients (indirectly, of course, through recorders), without CDI lookup. The Redis client doesn't do that, but we could probably add it. This would, on the other hand, require moving the build items to a `deployment-spi` kind of artifact. I'll explore more around that.

> I've played video games with 2.5K ;) In many cases I don't need my session data to be encrypted, just signed. Things like username, or even JWT token, don't need to be encrypted.

I can imagine even a simple cookie-based flash scope hitting the 2.5K boundary relatively easily if you use it to store a form with modest number of fields and validation messages. But then, I don't _really_ mind, adding a cookie-based session store would be easy. The only actual problem I can see is that the existing implementation in Vert.x Web does no size validation, and I'm not even sure what we could possibly do when we detect the limit was reached :shrug:

Ladicek commented:
I don't know to be honest. I do know that the Vert.x Web cookie-based session store doesn't do that out of the box :-)

FroMage commented:
Let's say we have "entity.name must have at least one character and at most 20 characters" as a validation constraint message (average size, I'd say), that's 71 chars. Most browsers have 4096 bytes for cookies. I'm not entirely sure as to the cookie value encoding, but Mozilla docs say US-ASCII, which is a pity, but 1 byte per char, so 71 bytes per message. So, in US-ASCII, we have store 57 error messages. That's a lot of elements for a form.

I'm not sure why we have to base64 the value. Or encrypt it. I've never had the need to hide session data from the user. Sign it, sure, but hide or base64 (except for text encoding, yeah), not sure. If there's a real use-case for encrypting session data, perhaps we have to wrap the Vert.x session API to make room for two cookies: one for non-encrypted text, one for encrypted/encoded?

Ladicek commented:
The [cookie-based session store](https://github.com/vert-x3/vertx-web/blob/master/vertx-web-session-stores/vertx-web-sstore-cookie/src/main/java/io/vertx/ext/web/sstore/cookie/impl/CookieSession.java) takes all the entries in the session, adds some metadata, serializes all that into a buffer, and encodes its content with Base64. That, and the Base64-encoded signature, is what gets into the cookie.

This is a generic implementation (though the set of possible types that you can put into the session is still fairly limited), a specialized one could be made more efficient I guess.

mkouba commented:
> > I don't want to support the cookie session store.
> 
> Why not? This is really the easiest to use for our users, because it requires no server dependency (unlike redis/etc) and allows for backend scaling unlike in-memory sessions.

Well, I think that the easiest and safest _default_ for most users is the in-memory session store. And that's what you get with no additional dependency so :shrug:.

Also for cookie store the sessions survive server restart/crashes which is not always what a user expects.

sberyozkin commented:
I'm not sure it is related, just in case, re the cookie session store:

>it is not encrypted, so you can't put sensitive information into it (and you probably shouldn't even if it was)

This is the default for OIDC and now it is encrypted by default, weak encryption will cause a warning, 32 chars key encryption won't be broken during the typical session lifetime, even the weaker key encryption one. Taking hold of non-encrypted OIDC tokens over HTTPS would be very hard on its own and ID token on its own would rarely have really sensitive data which would be fetched directly from OIDC server via UserInfo, while the Refresh token can't be used with knowing the client secret owned by Quarkus. Users don't need to know anything about clustering, Redis, Infinispan.

Having said that, I do support having such alternatives for OIDC users who no matter what will not trust the cookie session store :-). This is exactly why we now have a DB OIDC token state manager and I do like the idea of letting users to plugin one of Vertx Web Session options too if they'd want it

geoand commented:
I don't really have anything to offer, other folks have done a great job reviewing!

Ladicek commented:
> > Maybe I could use [this](https://quarkus.io/guides/conditional-extension-dependencies#declaring-a-conditional-dependency-in-the-quarkus-extension-descriptor) to avoid the dependency circularity? I didn't find any actual usage of that, so I'm not sure how that's supposed to work. I'll try, I guess...
> 
> I think you should be able to do that. Checking for a condition of Maven optional dependencies is a convenience, since, I was thinking at the time, typically they would be optional dependencies anyway.

This unfortunately doesn't seem to work. I didn't look too deep, but `mvn clean quarkus:dev` seems stuck in this area:

```
"main" #1 prio=5 os_prio=0 cpu=174000,94ms elapsed=175,51s tid=0x00007ff80c017800 nid=0x4a44 runnable  [0x00007ff8136ae000]
   java.lang.Thread.State: RUNNABLE
        at java.util.HashMap.putVal(java.base@11.0.19/HashMap.java:631)
        at java.util.HashMap.put(java.base@11.0.19/HashMap.java:608)
        at io.quarkus.bootstrap.resolver.maven.MavenArtifactResolver.newCollectManagedRequest(MavenArtifactResolver.java:325)
        at io.quarkus.bootstrap.resolver.maven.MavenArtifactResolver.collectManagedDependencies(MavenArtifactResolver.java:308)
        at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyTreeResolver.collectDependencies(ApplicationDependencyTreeResolver.java:592)
        at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyTreeResolver$ConditionalDependency.activate(ApplicationDependencyTreeResolver.java:757)
        at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyTreeResolver.resolve(ApplicationDependencyTreeResolver.java:183)
        at io.quarkus.bootstrap.resolver.BootstrapAppModelResolver.buildAppModel(BootstrapAppModelResolver.java:321)
        at io.quarkus.bootstrap.resolver.BootstrapAppModelResolver.doResolveModel(BootstrapAppModelResolver.java:286)
        at io.quarkus.bootstrap.resolver.BootstrapAppModelResolver.resolveManagedModel(BootstrapAppModelResolver.java:165)
        at io.quarkus.maven.QuarkusBootstrapProvider$QuarkusMavenAppBootstrap.doBootstrap(QuarkusBootstrapProvider.java:234)
        at io.quarkus.maven.QuarkusBootstrapProvider$QuarkusMavenAppBootstrap.bootstrapApplication(QuarkusBootstrapProvider.java:348)
        at io.quarkus.maven.QuarkusBootstrapProvider.bootstrapApplication(QuarkusBootstrapProvider.java:128)
        at io.quarkus.maven.QuarkusBootstrapMojo.bootstrapApplication(QuarkusBootstrapMojo.java:296)
        at io.quarkus.maven.GenerateCodeMojo.generateCode(GenerateCodeMojo.java:79)
        at io.quarkus.maven.GenerateCodeMojo.doExecute(GenerateCodeMojo.java:54)
        at io.quarkus.maven.QuarkusBootstrapMojo.execute(QuarkusBootstrapMojo.java:171)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
        at org.twdata.maven.mojoexecutor.MojoExecutor.executeMojo(MojoExecutor.java:120)
        at io.quarkus.maven.DevMojo.executeGoal(DevMojo.java:697)
        at io.quarkus.maven.DevMojo.handleAutoCompile(DevMojo.java:650)
        at io.quarkus.maven.DevMojo.execute(DevMojo.java:399)
        at ...
```

I would guess the circular dependency (which is invisible to Maven) leads to an infinite loop in the dependency resolution code anyway.

I'll investigate the build item based approach that doesn't require CDI lookups and hence doesn't require circular dependencies.

aloubyansky commented:
@Ladicek could you create a branch for me that results in that failure? I'd be interested in checking what's going on there. Thanks.

Ladicek commented:
Sure @aloubyansky, I pushed a branch here: https://github.com/Ladicek/quarkus/commits/vertx-web-sessions-circularity To test, you can build Quarkus from that branch, checkout https://github.com/Ladicek/quarkus-tinies, go to the `web-sessions` directory and run `mvn clean quarkus:dev`.

Ladicek committed:
Infinispan Client: add runtime SPI and deployment SPI modules

Some constants from the Infinispan Client runtime module are moved
to the runtime SPI module, so that they can be used from other modules
that cannot directly depend on the Infinispan Client extension.

Two build items, `InfinispanClientBuildItem` and `InfinispanClientNameBuildItem`,
are moved from the Infinispan Client deployment module to the deployment SPI
module. This is technically a breaking change, because they are also moved
to a different package, but these build items don't seem to be used anywhere
outside of Quarkus, so it should be safe.

Ladicek committed:
Redis Client: add runtime SPI and deployment SPI modules

Some constants from the Redis Client runtime module are moved
to the runtime SPI module, so that they can be used from other modules
that cannot directly depend on the Redis Client extension.

One build item, `RequestedRedisClientBuildItem`, is moved from the Redis Client
deployment module to the deployment SPI module. This is technically a breaking
change, because it is also moved to a different package, but this build item
doesn't seem to be used anywhere outside of Quarkus, so it should be safe.

Further, one new build item, `RedisClientBuildItem`, is added. It provides
runtime access to the Redis clients (in the Mutiny variant) without having
to perform a CDI lookup.

Ladicek committed:
Add support for Vert.x Web sessions

Ladicek commented:
Refactoring to avoid CDI lookups and using build items is apparently possible. Conditional extension dependencies can be used afterwards (where I use the optional dependency syntax, like all other Quarkus extensions that use this feature).

The first 2 commits now contain the necessary changes in the Infinispan Client and Redis Client extensions: introduction of a `deployment-spi` module, moving the necessary build items there (and creating the necessary build item in one case), adjusting the existing consumers. They are technically breaking changes, because the build items are moved to a different package, but there doesn't seem to be any consumer outside of Quarkus, so it should be safe.

The 3rd commit contains the implementation of sessions and is still marked as WIP. Next, I will try to simplify it, because the way how build-time and runtime code interact is fairly complex (and I hope that complexity can be reduced). Then, tests and documentation.

aloubyansky commented:
Just FYI, I see what the issue is with the conditional dependencies. I'll look into fixing it. Once we get the CVE fixes sorted out.

sberyozkin commented:
Hi @Ladicek as discussed above, please check how OIDC TokenStateManager can delegate to one of the Web Session options supporting the server side state, IMHO it is worth tuning the session API for OIDC be able to use it

Ladicek committed:
Infinispan Client: improve dev services logging

Ladicek commented:
I believe this is ready now.

Ladicek commented:
The failure in `TrustedXForwarderProxiesUnknownHostnameFailureTest.testHeadersAreIgnored` shows an actual bug in https://github.com/quarkusio/quarkus/blob/7a3fd9462e8c744f1e3420256557d65fc12a116a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedProxyHandler.java#L82-L83

Per the `io.vertx.core.dns.DnsClient.lookup()` javadoc, the lookup may succeed with a `null` value:

```java
* @param handler  the {@link io.vertx.core.Handler} to notify with the {@link io.vertx.core.AsyncResult}.
*                 The handler will get notified with the resolved address if a record was found. If non was found it
*                 will get notifed with {@code null}. If an error accours it will get failed.
```

 which the code linked above ignores. I believe we can treat the `null` value exactly as we treat a NXDOMAIN error. In fact, I think most DNS servers will just respond with NXDOMAIN, but for some reason, in the CI environment, we can get a successful `null` result, as evidenced by:

```
2023-10-24 18:07:15,785 ERROR [io.qua.ver.cor.run.VertxCoreRecorder] (vert.x-eventloop-thread-0) Uncaught exception received by Vert.x: java.lang.NullPointerException: Cannot invoke "String.indexOf(int)" because "address" is null
	at org.wildfly.common.net.Inet.parseInetAddress(Inet.java:629)
	at org.wildfly.common.net.Inet.parseInetAddress(Inet.java:616)
	at io.quarkus.vertx.http.runtime.ForwardedProxyHandler$1.handle(ForwardedProxyHandler.java:83)
	at io.quarkus.vertx.http.runtime.ForwardedProxyHandler$1.handle(ForwardedProxyHandler.java:79)
	at ...
```

I'll submit another PR with a fix.

Ladicek commented:
Submitted #36682

karesti commented:
@Ladicek from Infinispan side everything looks good! if tests pass correctly, we should not experience any breaking change. thanks!!

cescoffier commented on docs/src/main/asciidoc/http-reference.adoc:
```diff
@@ -546,7 +546,135 @@ link:https://undertow.io/undertow-docs/undertow-docs-2.0.0/index.html#predicates
+* `java.lang.String`
+* `byte[]`
+* `io.vertx.core.buffer.Buffer`
+* implementations of `io.vertx.core.shareddata.ClusterSerializable`
```
I would omit this one, as we do not really support clustering

Ladicek replied:
Done.

cescoffier commented on docs/src/main/asciidoc/http-reference.adoc:
```diff
@@ -546,7 +546,135 @@ link:https://undertow.io/undertow-docs/undertow-docs-2.0.0/index.html#predicates
+* `byte[]`
+* `io.vertx.core.buffer.Buffer`
+* implementations of `io.vertx.core.shareddata.ClusterSerializable`
+
```
What about JsonObject and JsonArray? Do you have to transform them into buffer before?

Ladicek replied:
`JsonObject` and `JsonArray` are actually `ClusterSerializable`, so they work just fine.

There's a few more classes in Vert.x that also implement `ClusterSerializable`, but those are not terribly interesting outside of Vert.x. I'll replace "implementations of `io.vertx.core.shareddata.ClusterSerializable`" with `JsonObject` and `JsonArray`.

Ladicek replied:
Done.

cescoffier commented on extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/devservices/InfinispanDevServiceProcessor.java:
```diff
@@ -189,7 +187,7 @@ private RunningDevService startContainer(String clientName, DockerStatusBuildIte
 
         String configPrefix = getConfigPrefix(clientName);
-        log.info("Config prefix " + configPrefix);
+        log.debug("Config prefix " + configPrefix);
```
debugf maybe?

Ladicek replied:
Will do.

Ladicek replied:
Done.

cescoffier reviewed (approved):
only minor comments.

Ladicek commented:
Tried to improve the documentation a little around the possibility of concurrent session access.

Ladicek commented:
Moved back to draft after conversation with @stuartwdouglas who indicated that Vert.x Web approach to concurrent session access (especially with external session stores) is not adequate enough for general purpose usage.

sberyozkin commented:
Hey @Ladicek, are you planning to restart working on this amazing PR :-) ?
